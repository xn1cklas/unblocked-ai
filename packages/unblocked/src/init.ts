import { defu } from 'defu';
// Password utilities removed for AI focus
import { createInternalAdapter, getMigrations } from './db';
import { getAITables } from './db/get-tables';
import { getAdapter } from './db/utils';
import { UnblockedError } from './error';
import type {
  Adapter,
  Models,
  SecondaryStorage,
  UnblockedOptions,
  UnblockedPlugin,
  User,
} from './types';
import type { LiteralUnion } from './types/helper';
// Social providers removed for AI focus
// OAuth removed for AI focus
import { generateId } from './utils';
import { DEFAULT_SECRET } from './utils/constants';
import { env, isProduction } from './utils/env';
// Cookies removed for AI focus
import { createLogger } from './utils/logger';
// Password utils removed for AI focus
import { getBaseURL } from './utils/url';

export const init = async (options: UnblockedOptions) => {
  const adapter = await getAdapter(options);
  const plugins = options.plugins || [];
  const internalPlugins = getInternalPlugins(options);
  const logger = createLogger(options.logger);

  const baseURL = getBaseURL(options.baseURL, options.basePath);

  const secret =
    options.secret || env.UNBLOCKED_SECRET || env.AUTH_SECRET || DEFAULT_SECRET;

  if (secret === DEFAULT_SECRET && isProduction) {
    logger.error(
      'You are using the default secret. Please set `UNBLOCKED_SECRET` in your environment variables or pass `secret` in your Unblocked config.'
    );
  }

  options = {
    ...options,
    secret,
    baseURL: baseURL ? new URL(baseURL).origin : '',
    basePath: options.basePath || '/api/unblocked',
    plugins: plugins.concat(internalPlugins),
  };
  // Cookies removed for AI focus
  const tables = getAITables(options);
  // Social providers removed for AI focus

  const generateIdFunc: UnblockedContext['generateId'] = ({ model, size }) => {
    if (typeof options.advanced?.generateId === 'function') {
      return options.advanced.generateId({ model, size });
    }
    if (typeof options?.advanced?.database?.generateId === 'function') {
      return options.advanced.database.generateId({ model, size });
    }
    return generateId(size);
  };

  const ctx: UnblockedContext = {
    appName: options.appName || 'Unblocked',
    // socialProviders removed for AI focus
    options,
    tables,
    trustedOrigins: getTrustedOrigins(options),
    baseURL: baseURL || '',
    // sessionConfig removed for AI focus
    secret,
    // rateLimit removed for AI focus - handled by AI providers
    // authCookies removed for AI focus
    logger,
    generateId: generateIdFunc,
    secondaryStorage: options.secondaryStorage,
    // Password handling removed for AI focus
    // session management removed for AI focus
    adapter,
    internalAdapter: createInternalAdapter(adapter, {
      options,
      hooks: options.databaseHooks ? [options.databaseHooks] : [],
      generateId: generateIdFunc,
    }),
    // createAuthCookie removed for AI focus
    async runMigrations() {
      //only run migrations if database is provided and it's not an adapter
      if (!options.database || 'updateMany' in options.database) {
        throw new UnblockedError(
          "Database is not provided or it's an adapter. Migrations are only supported with a database instance."
        );
      }
      const { runMigrations } = await getMigrations(options);
      await runMigrations();
    },
    // AI-specific properties
    getUser: async (request: Request) => {
      if (!(options.user && options.user.getUser)) {
        return null;
      }
      const result = await options.user.getUser(request);
      return result;
    },
  };
  const { context } = runPluginInit(ctx);
  return context;
};

export type UnblockedContext = {
  options: UnblockedOptions;
  appName: string;
  baseURL: string;
  trustedOrigins: string[];
  // session management removed for AI focus
  // socialProviders removed for AI focus
  // authCookies removed for AI focus
  logger: ReturnType<typeof createLogger>;
  // rateLimit removed for AI focus - handled by AI providers
  adapter: Adapter;
  internalAdapter: ReturnType<typeof createInternalAdapter>;
  // createAuthCookie removed for AI focus
  secret: string;
  // sessionConfig removed for AI focus
  generateId: (options: {
    model: LiteralUnion<Models, string>;
    size?: number;
  }) => string;
  secondaryStorage: SecondaryStorage | undefined;
  // password utilities removed for AI focus
  tables: ReturnType<typeof getAITables>;
  runMigrations: () => Promise<void>;
  // AI-specific properties
  getUser?: (request: Request) => Promise<User | null>;
  // Allow plugin extensions
  [key: string]: any;
};

function runPluginInit(ctx: UnblockedContext) {
  let options = ctx.options;
  const plugins = options.plugins || [];
  let context: UnblockedContext = ctx;
  const dbHooks: UnblockedOptions['databaseHooks'][] = [];
  for (const plugin of plugins) {
    if (plugin.init) {
      const result = plugin.init(context);
      if (typeof result === 'object') {
        if (result.options) {
          const { databaseHooks, ...restOpts } = result.options;
          if (databaseHooks) {
            dbHooks.push(databaseHooks);
          }
          options = defu(options, restOpts);
        }
        if (result.context) {
          context = {
            ...context,
            ...(result.context as Partial<UnblockedContext>),
          };
        }
      }
    }
  }
  // Add the global database hooks last
  dbHooks.push(options.databaseHooks);
  context.internalAdapter = createInternalAdapter(ctx.adapter, {
    options,
    hooks: dbHooks.filter((u) => u !== undefined),
    generateId: ctx.generateId,
  });
  context.options = options;
  return { context };
}

function getInternalPlugins(options: UnblockedOptions) {
  const plugins: UnblockedPlugin[] = [];
  // crossSubDomainCookies removed for AI focus
  return plugins;
}

function getTrustedOrigins(options: UnblockedOptions) {
  const baseURL = getBaseURL(options.baseURL, options.basePath);
  if (!baseURL) {
    return [];
  }
  const trustedOrigins = [new URL(baseURL).origin];
  if (options.trustedOrigins && Array.isArray(options.trustedOrigins)) {
    trustedOrigins.push(...options.trustedOrigins);
  }
  const envTrustedOrigins = env.UNBLOCKED_TRUSTED_ORIGINS;
  if (envTrustedOrigins) {
    trustedOrigins.push(...envTrustedOrigins.split(','));
  }
  if (trustedOrigins.filter((x) => !x).length) {
    throw new UnblockedError(
      'A provided trusted origin is invalid, make sure your trusted origins list is properly defined.'
    );
  }
  return trustedOrigins;
}
