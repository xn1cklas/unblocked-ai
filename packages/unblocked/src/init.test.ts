import Database from 'better-sqlite3';
import { describe, expect, it, vi } from 'vitest';
import { init } from './init';

// Auth client imports removed for AI focus
// import { getTestInstance } from "./test-utils/test-instance";

describe('init', async () => {
  const database = new Database(':memory:');

  it('should match config', async () => {
    const res = await init({
      baseURL: 'http://localhost:3000',
      database,
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('should infer BASE_URL from env', async () => {
    vi.stubEnv('UNBLOCKED_URL', 'http://localhost:5147');
    const res = await init({
      database,
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
    });
    expect(res.options.baseURL).toBe('http://localhost:5147');
    expect(res.baseURL).toBe('http://localhost:5147/api/unblocked');
    vi.unstubAllEnvs();
  });

  it('should respect base path', async () => {
    const res = await init({
      database,
      basePath: '/custom-path',
      baseURL: 'http://localhost:5147',
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
    });
    expect(res.baseURL).toBe('http://localhost:5147/custom-path');
  });

  it('should execute plugins init', async () => {
    const newBaseURL = 'http://test.test';
    const res = await init({
      baseURL: 'http://localhost:3000',
      database,
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
      plugins: [
        {
          id: 'test',
          init: () => {
            return {
              context: {
                baseURL: newBaseURL,
              },
            };
          },
        },
      ],
    });
    expect(res.baseURL).toBe(newBaseURL);
  });

  it('should work with custom path', async () => {
    const customPath = '/custom-path';
    const ctx = await init({
      database,
      basePath: customPath,
      baseURL: 'http://localhost:3000',
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
    });
    expect(ctx.baseURL).toBe(`http://localhost:3000${customPath}`);
    // Client test removed - requires handler setup
  });

  it('should allow plugins to set config values', async () => {
    const ctx = await init({
      database,
      baseURL: 'http://localhost:3000',
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
      plugins: [
        {
          id: 'test-plugin',
          init(ctx) {
            return {
              context: ctx,
              options: {
                appName: 'Custom AI App',
              },
            };
          },
        },
      ],
    });
    expect(ctx.options.appName).toBe('Custom AI App');
  });

  it('should not allow plugins to set config values if they are set in the main config', async () => {
    const ctx = await init({
      database,
      baseURL: 'http://localhost:3000',
      appName: 'Main AI App',
      user: {
        getUser: () => ({ id: 'test-user', name: 'Test User' }),
      },
      plugins: [
        {
          id: 'test-plugin',
          init(ctx) {
            return {
              context: ctx,
              options: {
                appName: 'Plugin AI App',
              },
            };
          },
        },
      ],
    });
    expect(ctx.options.appName).toBe('Main AI App');
  });
});
