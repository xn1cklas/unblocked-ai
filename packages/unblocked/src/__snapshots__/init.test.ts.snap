// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`init > should match config 1`] = `
{
  "adapter": {
    "count": [Function],
    "create": [Function],
    "createSchema": undefined,
    "delete": [Function],
    "deleteMany": [Function],
    "findMany": [Function],
    "findOne": [Function],
    "id": "kysely",
    "options": {
      "adapterConfig": {
        "adapterId": "kysely",
        "adapterName": "Kysely Adapter",
        "debugLogs": false,
        "supportsBooleans": false,
        "supportsDates": false,
        "supportsJSON": false,
        "supportsNumericIds": true,
        "usePlural": undefined,
      },
      "debugLogs": false,
      "type": "sqlite",
    },
    "update": [Function],
    "updateMany": [Function],
  },
  "appName": "Unblocked",
  "baseURL": "http://localhost:3000/api/unblocked",
  "generateId": [Function],
  "getUser": [Function],
  "internalAdapter": {
    "createChat": [Function],
    "createDocument": [Function],
    "createMessage": [Function],
    "createRateLimit": [Function],
    "createStream": [Function],
    "createSuggestion": [Function],
    "createVote": [Function],
    "deleteChat": [Function],
    "deleteDocument": [Function],
    "deleteMessage": [Function],
    "deleteRateLimit": [Function],
    "deleteStream": [Function],
    "deleteVote": [Function],
    "findChat": [Function],
    "findChats": [Function],
    "findChatsByUser": [Function],
    "findDocument": [Function],
    "findDocuments": [Function],
    "findDocumentsByUser": [Function],
    "findMessage": [Function],
    "findMessages": [Function],
    "findMessagesByChat": [Function],
    "findRateLimit": [Function],
    "findStream": [Function],
    "findSuggestion": [Function],
    "findSuggestionsByDocument": [Function],
    "findVote": [Function],
    "findVotesByMessage": [Function],
    "updateChat": [Function],
    "updateDocument": [Function],
    "updateMessage": [Function],
    "updateStream": [Function],
    "updateSuggestion": [Function],
    "updateVote": [Function],
  },
  "logger": {
    "debug": [Function],
    "error": [Function],
    "info": [Function],
    "success": [Function],
    "warn": [Function],
  },
  "options": {
    "basePath": "/api/unblocked",
    "baseURL": "http://localhost:3000",
    "database": Database {
      "inTransaction": false,
      "memory": true,
      "name": ":memory:",
      "open": true,
      "readonly": false,
    },
    "plugins": [],
    "secret": "unblocked-secret-123456789",
    "user": {
      "getUser": [Function],
    },
  },
  "runMigrations": [Function],
  "secondaryStorage": undefined,
  "secret": "unblocked-secret-123456789",
  "tables": {
    "chat": {
      "fields": {
        "createdAt": {
          "defaultValue": [Function],
          "required": true,
          "type": "date",
        },
        "title": {
          "required": true,
          "type": "string",
        },
        "userId": {
          "required": true,
          "type": "string",
        },
        "visibility": {
          "defaultValue": "private",
          "required": true,
          "type": [
            "public",
            "private",
          ],
        },
      },
      "modelName": "Chat",
      "order": 1,
    },
    "document": {
      "fields": {
        "content": {
          "required": false,
          "type": "string",
        },
        "createdAt": {
          "defaultValue": [Function],
          "required": true,
          "type": "date",
        },
        "kind": {
          "defaultValue": "text",
          "required": true,
          "type": [
            "text",
            "code",
            "image",
            "sheet",
          ],
        },
        "title": {
          "required": true,
          "type": "string",
        },
        "userId": {
          "required": true,
          "type": "string",
        },
      },
      "modelName": "Document",
      "order": 4,
    },
    "message": {
      "fields": {
        "attachments": {
          "required": true,
          "type": "string",
        },
        "chatId": {
          "references": {
            "field": "id",
            "model": "Chat",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
        "createdAt": {
          "defaultValue": [Function],
          "required": true,
          "type": "date",
        },
        "parts": {
          "required": true,
          "type": "string",
        },
        "role": {
          "required": true,
          "type": "string",
        },
      },
      "modelName": "Message",
      "order": 2,
    },
    "stream": {
      "fields": {
        "chatId": {
          "references": {
            "field": "id",
            "model": "Chat",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
        "createdAt": {
          "defaultValue": [Function],
          "required": true,
          "type": "date",
        },
      },
      "modelName": "Stream",
      "order": 6,
    },
    "suggestion": {
      "fields": {
        "createdAt": {
          "defaultValue": [Function],
          "required": true,
          "type": "date",
        },
        "description": {
          "required": false,
          "type": "string",
        },
        "documentCreatedAt": {
          "required": true,
          "type": "date",
        },
        "documentId": {
          "required": true,
          "type": "string",
        },
        "isResolved": {
          "defaultValue": false,
          "required": true,
          "type": "boolean",
        },
        "originalText": {
          "required": true,
          "type": "string",
        },
        "suggestedText": {
          "required": true,
          "type": "string",
        },
        "userId": {
          "required": true,
          "type": "string",
        },
      },
      "modelName": "Suggestion",
      "order": 5,
    },
    "vote": {
      "fields": {
        "chatId": {
          "references": {
            "field": "id",
            "model": "Chat",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
        "isUpvoted": {
          "required": true,
          "type": "boolean",
        },
        "messageId": {
          "references": {
            "field": "id",
            "model": "Message",
            "onDelete": "cascade",
          },
          "required": true,
          "type": "string",
        },
      },
      "modelName": "Vote",
      "order": 3,
    },
  },
  "trustedOrigins": [
    "http://localhost:3000",
  ],
}
`;
